compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirContractChecker.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.contracts.FirResolvedContractDescription
+import org.jetbrains.kotlin.fir.declarations.FirContractDescriptionOwner
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.expressions.FirLoop
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.expressions.FirWhenExpression
+
+private const val UNEXPECTED_CONSTRUCTION = "Unexpected construction in contract description"
+
+object FirContractChecker : FirFunctionChecker() {
+    override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration !is FirContractDescriptionOwner ||
+            declaration.contractDescription !is FirResolvedContractDescription
+        ) {
+            return
+        }
+
+        for (statement in (declaration.contractDescription as FirResolvedContractDescription).unresolvedEffects) {
```

Comment:
Just curious, what are these `unresolvedEffects` supposed to be? Is it expected to have any at all? What is an example of a valid statement that is not an effect? The name seems to imply to me that there shouldn't be any at all. Disclaimer: This is new to me. 🙂

Reply:
It will have any statements that `ConeEffectExtractor` cannot extract effects. I added a comment in code, just in case.